runtime/sema/check_transaction_declaration.go

```diff
@@ -248,36 +303,175 @@ func (checker *Checker) declareTransactionDeclaration(declaration *ast.Transacti
 		transactionType.Parameters = checker.parameters(declaration.ParameterList)
 	}
 
+	members, fields, origins := checker.compositeFieldMembersAndOrigins(
+		transactionType,
+		declaration.DeclarationKind(),
+		declaration.Fields,
+	)
+
+	transactionType.Members = members
+	transactionType.Fields = fields
+	if checker.PositionInfo != nil {
+		checker.PositionInfo.recordMemberOrigins(transactionType, origins)
+	}
+
+	if declaration.Prepare != nil {
+		parameterList := declaration.Prepare.FunctionDeclaration.ParameterList
+		transactionType.PrepareParameters = checker.parameters(parameterList)
+	}
+
+	var fieldDeclarationsByIdentifier map[string]*ast.FieldDeclaration
+	fieldDeclarationByIdentifier := func(identifier string) *ast.FieldDeclaration {
+		if fieldDeclarationsByIdentifier == nil {
+			fieldDeclarationsByIdentifier = make(map[string]*ast.FieldDeclaration, len(declaration.Fields))
+			for _, field := range declaration.Fields {
+				fieldDeclarationsByIdentifier[field.Identifier.Identifier] = field
+			}
+		}
+		return fieldDeclarationsByIdentifier[identifier]
+	}
+
+	roles := &orderedmap.OrderedMap[string, *TransactionRoleType]{}
+	for _, roleDeclaration := range declaration.Roles {
+		transactionRoleType := checker.transactionRoleType(roleDeclaration)
+		checker.Elaboration.SetTransactionRoleDeclarationType(roleDeclaration, transactionRoleType)
+
+		// Ensure roles are not duplicated
+		roleName := roleDeclaration.Identifier.Identifier
+		if _, ok := roles.Get(roleName); ok {
+			checker.report(
+				&DuplicateTransactionRoleError{
+					Name: roleName,
+					Range: ast.NewRangeFromPositioned(
+						checker.memoryGauge,
+						roleDeclaration.Identifier,
+					),
+				},
+			)
+		} else {
+			roles.Set(roleName, transactionRoleType)
+		}
+
+		// Ensure roles and fields do not clash
+		if _, ok := members.Get(roleName); ok {
+			field := fieldDeclarationByIdentifier(roleName)
+			if field != nil {
+				checker.report(
+					&TransactionRoleWithFieldNameError{
+						FieldIdentifier: field.Identifier,
+						Range: ast.NewRangeFromPositioned(
+							checker.memoryGauge,
+							roleDeclaration.Identifier,
+						),
+					},
+				)
+			}
+		} else {
+			members.Set(
+				roleName,
+				&Member{
+					ContainerType:   transactionType,
+					Identifier:      roleDeclaration.Identifier,
+					DeclarationKind: common.DeclarationKindTransactionRole,
+					VariableKind:    ast.VariableKindConstant,
+					TypeAnnotation:  NewTypeAnnotation(transactionRoleType),
+					DocString:       roleDeclaration.DocString,
+				},
+			)
+		}
+	}
+	transactionType.Roles = roles
```

Comment:
Because it is later needed to iterate over all the rules and check if their resource-kinded fields were invalidated. It is true that this information is also available in `transactionType.Members`, but that also contains fields, so needs filtering. I'll refactor

Reply:
Refactored in 67c69228316b47dde4f8599674a319bc9154ba99